### PR TITLE
ASR patches for v1.0.0

### PR DESCRIPTION
**Pull request by titu1994**
# Bugfixes
- Correct WER calculation for RNNT by resetting the metric after each computation.
- Prevent numba from performing a sync across PT cuda stream simply to initialize a cost vector.
- Correctly update the trainer value using `set_trainer`
- Adds .nemo and .ckpt files to dockerignore

# Additions
- Adds RNNT global flags to enable detached gradient normalization of encoder, decoder and joint models
- Adds `weights_init_scale` and `hidden_hidden_bias_scale` values (default to original values) to attempt reduction in variance between runs
- Add `tds_uniform` and `tds_normal` initialization modes for Jasper based models.

### CODE REVIEW

**Comment by titu1994** [comment]
@okuchaiev fyi regarding the dockerfile ignore and correctly updating the self.trainer inside set_trainer.

File: nemo/collections/asr/parts/jasper.py [hunk 1]
```diff
@@ -36,6 +37,28 @@
 jasper_activations = {"hardtanh": nn.Hardtanh, "relu": nn.ReLU, "selu": nn.SELU, "swish": Swish}
 
 
+def tds_uniform_(tensor, mode='fan_in'):
+    """
```
Comment (VahidooX) [inline nemo/collections/asr/parts/jasper.py:41]: Should they get dropped?

File: nemo/collections/asr/parts/jasper.py [hunk 2]
```diff
@@ -36,6 +37,28 @@
...
+    fan = _calculate_correct_fan(tensor, mode)
+    gain = 2.0
+    std = gain / math.sqrt(fan)  # sqrt(4.0 / fan_in)
+    bound = std  # Calculate uniform bounds from standard deviation
+    with torch.no_grad():
+        return tensor.uniform_(-bound, bound)
+
+
+def tds_normal_(tensor, mode='fan_in'):
+    """
```
Comment (VahidooX) [inline nemo/collections/asr/parts/jasper.py:52]: Should they get dropped?

File: nemo/collections/asr/models/rnnt_models.py [hunk 3]
```diff
@@ -120,6 +123,32 @@ def __init__(self, cfg: DictConfig, trainer: Trainer = None):
             self._optim_variational_noise_std = 0
             self._optim_variational_noise_start = 0
 
+        # Setup normalized gradients for model
+        if 'normalize_joint' in self.cfg:
```
Comment (VahidooX) [inline nemo/collections/asr/models/rnnt_models.py:127]: Can they be done like this?
self._normalize_joint = self.cfg.get(normalize_joint, False)

File: nemo/collections/asr/models/rnnt_models.py [hunk 4]
```diff
@@ -120,6 +123,32 @@ def __init__(self, cfg: DictConfig, trainer: Trainer = None):
...
+            self._optim_normalize_joint = self.cfg.normalize_joint
+            self._optim_normalize_txu = None
+        else:
+            self._optim_normalize_joint = False
+            self._optim_normalize_txu = None
+
+        # Setup normalized encoder norm for model
+        if 'normalize_encoder_norm' in self.cfg:
```
Comment (VahidooX) [inline nemo/collections/asr/models/rnnt_models.py:135]: These four parameters, if they are similar and related, how about adding a new section like optim_normalization and add them under that? Just a suggestion to organize them. It can also make the names smaller too.

File: nemo/collections/asr/models/rnnt_models.py [hunk 5]
```diff
@@ -120,6 +123,32 @@ def __init__(self, cfg: DictConfig, trainer: Trainer = None):
...
+            self._optim_normalize_encoder_norm = self.cfg.normalize_encoder_norm
+        else:
+            self._optim_normalize_encoder_norm = False
+
+        # Setup normalized decoder norm for model
+        if 'normalize_decoder_norm' in self.cfg:
```
Comment (VahidooX) [inline nemo/collections/asr/models/rnnt_models.py:141]: Another suggestion is that as all have _norm except one, you may drop that but make the normalize_joint to normalize_joint_txu?

File: nemo/collections/asr/modules/rnnt.py [hunk 6]
```diff
@@ -245,7 +253,18 @@ def predict(
...
 
-    def _predict(self, vocab_size, pred_n_hidden, pred_rnn_layers, forget_gate_bias, t_max, norm, dropout):
+    def _predict(
+        self,
+        vocab_size,
+        pred_n_hidden,
+        pred_rnn_layers,
+        forget_gate_bias,
+        t_max,
+        norm,
```
Comment (VahidooX) [inline nemo/collections/asr/modules/rnnt.py:263]: Descriptions of the newe params are missing.

File: nemo/collections/asr/parts/jasper.py [hunk 7]
```diff
@@ -36,6 +37,28 @@
...
+    """
+    fan = _calculate_correct_fan(tensor, mode)
+    gain = 2.0
```
Comment (VahidooX) [inline nemo/collections/asr/parts/jasper.py:55]: Can these initialization be useful for others? If yes, you may move them somewhere where is shared?

File: nemo/collections/asr/parts/numba/rnnt_loss/rnnt.py [hunk 8]
```diff
@@ -180,7 +180,7 @@ def rnnt_loss_gpu(
     acts, acts_shape = rnnt_helper.flatten_tensor(acts)
 
     ### REPRESENT THE CUDA ARRAY INTERFACE OF COSTS VECTOR ###
-    costs_repr = cuda.as_cuda_array(costs)  # NO COPY OF DATA, JUST CHANGE REPRESENTATION
+    costs_repr = cuda.as_cuda_array(costs, sync=False)  # NO COPY OF DATA, JUST CHANGE REPRESENTATION
```
Comment (VahidooX) [inline nemo/collections/asr/parts/numba/rnnt_loss/rnnt.py:183]: Is it OK to not sync it?

File: nemo/collections/common/parts/rnn.py [hunk 9]
```diff
@@ -188,10 +204,15 @@ def __init__(
                     bias.data[hidden_size : 2 * hidden_size].fill_(forget_gate_bias)
                 if "bias_hh" in name:
                     bias = getattr(self.lstm, name)
-                    bias.data[hidden_size : 2 * hidden_size].fill_(0)
+                    # bias.data[hidden_size : 2 * hidden_size].fill_(0)
```
Comment (VahidooX) [inline nemo/collections/common/parts/rnn.py:207]: Drop the comment?

File: nemo/collections/common/parts/rnn.py [hunk 10]
```diff
@@ -188,10 +204,15 @@ def __init__(
...
+                    bias.data[hidden_size : 2 * hidden_size] *= float(hidden_hidden_bias_scale)
 
         self.dropout = torch.nn.Dropout(dropout) if dropout else None
 
+        for name, v in self.named_parameters():
```
Comment (VahidooX) [inline nemo/collections/common/parts/rnn.py:212]: How about to check and if weights_init_scale is 1.0 then skip this part?

File: nemo/core/classes/modelPT.py [hunk 11]
```diff
@@ -1177,6 +1177,7 @@ def set_trainer(self, trainer: Trainer):
         Args:
             trainer: PyTorch Lightning Trainer object.
         """
+        self.trainer = trainer
```
Comment (VahidooX) [inline nemo/core/classes/modelPT.py:1180]: Why do we keep both _trianer and trianer?

File: nemo/collections/common/parts/rnn.py [hunk 12]
```diff
@@ -138,6 +152,8 @@ def __init__(
         dropout: Optional[float],
         forget_gate_bias: Optional[float],
         t_max: Optional[int] = None,
+        weights_init_scale: float = 1.0,
```
Comment (VahidooX) [inline nemo/collections/common/parts/rnn.py:155]: Descriptions of the new params are missing.

File: nemo/collections/common/parts/rnn.py [hunk 13]
```diff
@@ -188,10 +204,15 @@ def __init__(
                     bias.data[hidden_size : 2 * hidden_size].fill_(forget_gate_bias)
                 if "bias_hh" in name:
                     bias = getattr(self.lstm, name)
-                    bias.data[hidden_size : 2 * hidden_size].fill_(0)
+                    # bias.data[hidden_size : 2 * hidden_size].fill_(0)
+                    bias.data[hidden_size : 2 * hidden_size] *= float(hidden_hidden_bias_scale)
```
Comment (VahidooX) [inline nemo/collections/common/parts/rnn.py:208]: How about to check and if hidden_hidden_bias_scale is 0.0 then skip this part?

Comment (VahidooX) [reply nemo/collections/common/parts/rnn.py:208]: Is it correct that you are multiplying by zero here as the default.?
 Just to make sure it is on purpose.

File: nemo/collections/asr/parts/jasper.py (hunk 1 above)
Comment (VahidooX) [inline nemo/collections/asr/parts/jasper.py:41]: Is there any references for these initialization techniques to mention here?

**Review by VahidooX — CHANGES_REQUESTED** [review]

File: nemo/collections/asr/models/rnnt_models.py (hunk 4 above)
Comment (titu1994) [reply nemo/collections/asr/models/rnnt_models.py]: Yes we can rename this function to `_setup_optim_normalization`

File: nemo/collections/asr/models/rnnt_models.py (hunk 5 above)
Comment (titu1994) [reply nemo/collections/asr/models/rnnt_models.py]: Good point. We also need docstrings here.

File: nemo/collections/asr/modules/rnnt.py (hunk 6 above)
Comment (titu1994) [reply nemo/collections/asr/modules/rnnt.py:263]: Good catch

File: nemo/collections/asr/parts/jasper.py (hunk 1 above)
Comment (titu1994) [reply nemo/collections/asr/parts/jasper.py:41]: Ah i forgot to fill this.

File: nemo/collections/asr/parts/jasper.py (hunk 7 above)
Comment (titu1994) [reply nemo/collections/asr/parts/jasper.py:79]: Hmm they are somewhat specific to Jasper based models right now (ill add a reference to the paper). We can move them to commons if anyone needs it in the future.

File: nemo/collections/asr/parts/numba/rnnt_loss/rnnt.py (hunk 8 above)
Comment (titu1994) [reply nemo/collections/asr/parts/numba/rnnt_loss/rnnt.py:183]: Yup tested this. Was the cause of a minor slow down since it would wait for the default pytorch stream to execute and sync before continuing, which was unnecessary since these kernels will just queue more stuff to that same stream.

File: nemo/collections/common/parts/rnn.py (hunk 12 above)
Comment (titu1994) [reply nemo/collections/common/parts/rnn.py:155]: Good catch.

File: nemo/collections/common/parts/rnn.py (hunk 13 above)
Comment (titu1994) [reply nemo/collections/common/parts/rnn.py:213]: It's done just once at init, so its fast enough. The default here is matching what was done in general prior to this commit. Mathematically it is equivalent.

File: nemo/collections/common/parts/rnn.py (hunk 10 above)
Comment (titu1994) [reply nemo/collections/common/parts/rnn.py:217]: Its done only once at init, should not cost much time.

File: nemo/core/classes/modelPT.py (hunk 11 above)
Comment (titu1994) [reply nemo/core/classes/modelPT.py:1180]: PTL 1.2.8+ refers to self.trainer in order to get values such as self.node_rank, self.global_rank etc. So it needs to be explicitly updated. The `_trainer` is for backward compat with older NeMo models.

File: nemo/collections/asr/parts/jasper.py (hunk 1 above)
Comment (titu1994) [reply nemo/collections/asr/parts/jasper.py:41]: Yup I forgot to add it in a hurry.

**Review by titu1994 — COMMENTED** [review]
Addressed comments

File: nemo/core/classes/modelPT.py (hunk 11 above)
Comment (VahidooX) [reply nemo/core/classes/modelPT.py:1180]: In that case, why haven't we change our codes to use _trainer instead of trainer? It is just a find and replace?

File: nemo/collections/asr/models/rnnt_models.py (hunk 4 above)
Comment (VahidooX) [reply nemo/collections/asr/models/rnnt_models.py]: I meant adding a section to the config file like this:
model:
    optim_normalization:
         normalize_encoder_norm: true
         normalize_joint:false
    ...

the current form is also OK, just a suggestion to make it more organized as they are at least 4 and maybe more in future may be added.

**Review by VahidooX — APPROVED** [review]

File: nemo/core/classes/modelPT.py (hunk 11 above)
Comment (titu1994) [reply nemo/core/classes/modelPT.py:1180]: They both are same thing and are practically the same reference. Newer code should just use self.trainer directly.

File: nemo/collections/asr/models/rnnt_models.py (hunk 4 above)
Comment (titu1994) [reply nemo/collections/asr/models/rnnt_models.py]: Yes that can be another option. Other than variational noise, I expect the other 3 to be rarely used. They only really apply to very specific combination of model and optimizer so we might refactor them into their own section if they every become more useful in general case
